# Association Rule Mining

## Importing Libraries

In [1]:
import pandas as pd
import numpy as np

In [2]:
from sklearn.preprocessing import OrdinalEncoder
from mlxtend.preprocessing import TransactionEncoder
from mlxtend.frequent_patterns import apriori
from mlxtend.frequent_patterns import association_rules

## Reading Data for arules 1 and Data for arules 2

In [3]:
arules1=pd.read_csv("C:/Data/Data for arules 1.csv")

In [4]:
arules2=pd.read_csv("C:/Data/Data for arules 2.csv")

## 1. Association Rule for Data for arules 1 :

In [5]:
arules1.head()

resid  familysz  chb18  cha18 borvita milo zandu dred dblue prtnex holik  \
0      1         3      1      0     yes  yes    no   no    no     no   yes   
1      2         6      0      2      no   no   yes   no   yes     no    no   
2      3         4      1      1      no   no    no   no    no    yes   yes   
3      4         4      0      2     yes  yes    no   no    no     no   yes   
4      5         3      1      0     yes  yes    no   no    no     no   yes   

  badya complan  
0    no      no  
1   yes      no  
2    no     yes  
3    no      no  
4    no     yes

### Removing first 4 columns

In [6]:
newarules1 = pd.DataFrame(arules1)
newarules1.drop(["resid","familysz","chb18","cha18"],axis=1,inplace=True)
newarules1.head()

borvita milo zandu dred dblue prtnex holik badya complan
0     yes  yes    no   no    no     no   yes    no      no
1      no   no   yes   no   yes     no    no   yes      no
2      no   no    no   no    no    yes   yes    no     yes
3     yes  yes    no   no    no     no   yes    no      no
4     yes  yes    no   no    no     no   yes    no     yes

### Ordinal Encoder () used to convert "Yes" and "No" to 1 and 0 and formed an array

In [8]:
# Converting the yes and no values to 1 and 0 respectively.
transformation = OrdinalEncoder()
newarules_1 = transformation.fit_transform(newarules1)
newarules_1

array([[1., 1., 0., 0., 0., 0., 1., 0., 0.],
       [0., 0., 1., 0., 1., 0., 0., 1., 0.],
       [0., 0., 0., 0., 0., 1., 1., 0., 1.],
       [1., 1., 0., 0., 0., 0., 1., 0., 0.],
       [1., 1., 0., 0., 0., 0., 1., 0., 1.],
       [0., 0., 0., 0., 0., 1., 1., 0., 1.],
       [0., 0., 1., 1., 1., 0., 0., 1., 0.],
       [0., 0., 1., 0., 1., 0., 0., 1., 0.],
       [0., 0., 0., 1., 1., 0., 0., 1., 0.],
       [1., 1., 0., 0., 0., 0., 1., 0., 0.],
       [1., 1., 0., 0., 0., 0., 1., 0., 0.],
       [1., 1., 0., 0., 0., 1., 1., 0., 1.],
       [0., 0., 0., 0., 0., 1., 1., 0., 1.],
       [0., 0., 0., 0., 0., 1., 1., 0., 1.],
       [1., 1., 0., 0., 0., 0., 1., 0., 0.],
       [0., 0., 0., 0., 0., 1., 1., 0., 1.],
       [1., 1., 0., 0., 0., 0., 1., 0., 0.],
       [0., 0., 1., 1., 1., 0., 0., 1., 0.],
       [1., 1., 0., 0., 0., 0., 1., 0., 1.],
       [1., 1., 0., 0., 0., 0., 1., 0., 0.],
       [1., 1., 0., 0., 0., 0., 1., 0., 0.],
       [1., 1., 0., 0., 0., 1., 1., 0., 1.],
       [0.

In [10]:
df=pd.DataFrame(newarules_1)
df

0    1    2    3    4    5    6    7    8
0   1.0  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0
1   0.0  0.0  1.0  0.0  1.0  0.0  0.0  1.0  0.0
2   0.0  0.0  0.0  0.0  0.0  1.0  1.0  0.0  1.0
3   1.0  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0
4   1.0  1.0  0.0  0.0  0.0  0.0  1.0  0.0  1.0
5   0.0  0.0  0.0  0.0  0.0  1.0  1.0  0.0  1.0
6   0.0  0.0  1.0  1.0  1.0  0.0  0.0  1.0  0.0
7   0.0  0.0  1.0  0.0  1.0  0.0  0.0  1.0  0.0
8   0.0  0.0  0.0  1.0  1.0  0.0  0.0  1.0  0.0
9   1.0  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0
10  1.0  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0
11  1.0  1.0  0.0  0.0  0.0  1.0  1.0  0.0  1.0
12  0.0  0.0  0.0  0.0  0.0  1.0  1.0  0.0  1.0
13  0.0  0.0  0.0  0.0  0.0  1.0  1.0  0.0  1.0
14  1.0  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0
15  0.0  0.0  0.0  0.0  0.0  1.0  1.0  0.0  1.0
16  1.0  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0
17  0.0  0.0  1.0  1.0  1.0  0.0  0.0  1.0  0.0
18  1.0  1.0  0.0  0.0  0.0  0.0  1.0  0.0  1.0
19  1.0  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0
20  1.0  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0
21  1.0  1.0  0.0  0.0  0.0  1.0  1.0  0.0  1.0
22  0.0  0.0  0.0  0.0  0.0  1.0  1.0  0.0  1.0
23  0.0  0.0  0.0  0.0  0.0  1.0  1.0  0.0  1.0
24  0.0  0.0  1.0  0.0  1.0  0.0  0.0  1.0  0.0
25  0.0  0.0  1.0  1.0  1.0  0.0  0.0  1.0  0.0
26  0.0  0.0  1.0  0.0  1.0  0.0  0.0  1.0  0.0
27  0.0  0.0  0.0  1.0  1.0  0.0  0.0  1.0  0.0
28  1.0  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0
29  0.0  0.0  0.0  0.0  0.0  1.0  1.0  0.0  1.0
30  1.0  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0
31  1.0  1.0  0.0  0.0  0.0  0.0  1.0  0.0  1.0
32  0.0  0.0  0.0  0.0  0.0  1.0  1.0  0.0  1.0
33  1.0  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0
34  1.0  1.0  0.0  0.0  0.0  0.0  1.0  0.0  0.0
35  1.0  1.0  0.0  0.0  0.0  1.0  1.0  0.0  1.0
36  0.0  0.0  0.0  0.0  0.0  1.0  1.0  0.0  1.0
37  0.0  0.0  0.0  0.0  0.0  1.0  1.0  0.0  1.0
38  0.0  0.0  1.0  0.0  1.0  0.0  0.0  1.0  0.0
39  0.0  0.0  1.0  1.0  1.0  0.0  0.0  1.0  0.0

### Creating dataframe of dataset from array above with column names

In [11]:
new_df1=df.rename(columns={0: "borvita", 1: "milo",2:'zandu',3:'dred',4:'dblue',5:'prtnex',6:'holik',7:'badya',8:'complan'})
new_df1

borvita  milo  zandu  dred  dblue  prtnex  holik  badya  complan
0       1.0   1.0    0.0   0.0    0.0     0.0    1.0    0.0      0.0
1       0.0   0.0    1.0   0.0    1.0     0.0    0.0    1.0      0.0
2       0.0   0.0    0.0   0.0    0.0     1.0    1.0    0.0      1.0
3       1.0   1.0    0.0   0.0    0.0     0.0    1.0    0.0      0.0
4       1.0   1.0    0.0   0.0    0.0     0.0    1.0    0.0      1.0
5       0.0   0.0    0.0   0.0    0.0     1.0    1.0    0.0      1.0
6       0.0   0.0    1.0   1.0    1.0     0.0    0.0    1.0      0.0
7       0.0   0.0    1.0   0.0    1.0     0.0    0.0    1.0      0.0
8       0.0   0.0    0.0   1.0    1.0     0.0    0.0    1.0      0.0
9       1.0   1.0    0.0   0.0    0.0     0.0    1.0    0.0      0.0
10      1.0   1.0    0.0   0.0    0.0     0.0    1.0    0.0      0.0
11      1.0   1.0    0.0   0.0    0.0     1.0    1.0    0.0      1.0
12      0.0   0.0    0.0   0.0    0.0     1.0    1.0    0.0      1.0
13      0.0   0.0    0.0   0.0    0.0     1.0    1.0    0.0      1.0
14      1.0   1.0    0.0   0.0    0.0     0.0    1.0    0.0      0.0
15      0.0   0.0    0.0   0.0    0.0     1.0    1.0    0.0      1.0
16      1.0   1.0    0.0   0.0    0.0     0.0    1.0    0.0      0.0
17      0.0   0.0    1.0   1.0    1.0     0.0    0.0    1.0      0.0
18      1.0   1.0    0.0   0.0    0.0     0.0    1.0    0.0      1.0
19      1.0   1.0    0.0   0.0    0.0     0.0    1.0    0.0      0.0
20      1.0   1.0    0.0   0.0    0.0     0.0    1.0    0.0      0.0
21      1.0   1.0    0.0   0.0    0.0     1.0    1.0    0.0      1.0
22      0.0   0.0    0.0   0.0    0.0     1.0    1.0    0.0      1.0
23      0.0   0.0    0.0   0.0    0.0     1.0    1.0    0.0      1.0
24      0.0   0.0    1.0   0.0    1.0     0.0    0.0    1.0      0.0
25      0.0   0.0    1.0   1.0    1.0     0.0    0.0    1.0      0.0
26      0.0   0.0    1.0   0.0    1.0     0.0    0.0    1.0      0.0
27      0.0   0.0    0.0   1.0    1.0     0.0    0.0    1.0      0.0
28      1.0   1.0    0.0   0.0    0.0     0.0    1.0    0.0      0.0
29      0.0   0.0    0.0   0.0    0.0     1.0    1.0    0.0      1.0
30      1.0   1.0    0.0   0.0    0.0     0.0    1.0    0.0      0.0
31      1.0   1.0    0.0   0.0    0.0     0.0    1.0    0.0      1.0
32      0.0   0.0    0.0   0.0    0.0     1.0    1.0    0.0      1.0
33      1.0   1.0    0.0   0.0    0.0     0.0    1.0    0.0      0.0
34      1.0   1.0    0.0   0.0    0.0     0.0    1.0    0.0      0.0
35      1.0   1.0    0.0   0.0    0.0     1.0    1.0    0.0      1.0
36      0.0   0.0    0.0   0.0    0.0     1.0    1.0    0.0      1.0
37      0.0   0.0    0.0   0.0    0.0     1.0    1.0    0.0      1.0
38      0.0   0.0    1.0   0.0    1.0     0.0    0.0    1.0      0.0
39      0.0   0.0    1.0   1.0    1.0     0.0    0.0    1.0      0.0

## Creating table for frequent items bought with minimum Support=30%
### Support refers to the popularity of a product out of total transactions. Here we have considered min support to be 30% and more

In [12]:
freq_items1=apriori(new_df1,min_support=0.3,use_colnames=True)
freq_items1

support                  itemsets
0     0.450                 (borvita)
1     0.450                    (milo)
2     0.350                  (prtnex)
3     0.725                   (holik)
4     0.425                 (complan)
5     0.450           (milo, borvita)
6     0.450          (holik, borvita)
7     0.450             (milo, holik)
8     0.350           (prtnex, holik)
9     0.350         (prtnex, complan)
10    0.425          (complan, holik)
11    0.450    (milo, holik, borvita)
12    0.350  (prtnex, complan, holik)

## Next Rule applied is for Confidence where minimum Confidence is 30%
### Confidence is the probability of buying product 2 if product 1 is bought. Below we have considered the rule where confidence is 30% and more. For below dataset most of the items ahve confidence level 100%.

In [13]:
rules=association_rules(freq_items1,metric="confidence",min_threshold=0.3)
rules

antecedents        consequents  antecedent support  \
0              (milo)          (borvita)               0.450   
1           (borvita)             (milo)               0.450   
2             (holik)          (borvita)               0.725   
3           (borvita)            (holik)               0.450   
4              (milo)            (holik)               0.450   
5             (holik)             (milo)               0.725   
6            (prtnex)            (holik)               0.350   
7             (holik)           (prtnex)               0.725   
8            (prtnex)          (complan)               0.350   
9           (complan)           (prtnex)               0.425   
10          (complan)            (holik)               0.425   
11            (holik)          (complan)               0.725   
12      (milo, holik)          (borvita)               0.450   
13    (milo, borvita)            (holik)               0.450   
14   (holik, borvita)             (milo)               0.450   
15             (milo)   (holik, borvita)               0.450   
16            (holik)    (milo, borvita)               0.725   
17          (borvita)      (milo, holik)               0.450   
18  (prtnex, complan)            (holik)               0.350   
19    (prtnex, holik)          (complan)               0.350   
20   (complan, holik)           (prtnex)               0.425   
21           (prtnex)   (complan, holik)               0.350   
22          (complan)    (prtnex, holik)               0.425   
23            (holik)  (prtnex, complan)               0.725   

    consequent support  support  confidence      lift  leverage  conviction  
0                0.450    0.450    1.000000  2.222222  0.247500         inf  
1                0.450    0.450    1.000000  2.222222  0.247500         inf  
2                0.450    0.450    0.620690  1.379310  0.123750    1.450000  
3                0.725    0.450    1.000000  1.379310  0.123750         inf  
4                0.725    0.450    1.000000  1.379310  0.123750         inf  
5                0.450    0.450    0.620690  1.379310  0.123750    1.450000  
6                0.725    0.350    1.000000  1.379310  0.096250         inf  
7                0.350    0.350    0.482759  1.379310  0.096250    1.256667  
8                0.425    0.350    1.000000  2.352941  0.201250         inf  
9                0.350    0.350    0.823529  2.352941  0.201250    3.683333  
10               0.725    0.425    1.000000  1.379310  0.116875         inf  
11               0.425    0.425    0.586207  1.379310  0.116875    1.389583  
12               0.450    0.450    1.000000  2.222222  0.247500         inf  
13               0.725    0.450    1.000000  1.379310  0.123750         inf  
14               0.450    0.450    1.000000  2.222222  0.247500         inf  
15               0.450    0.450    1.000000  2.222222  0.247500         inf  
16               0.450    0.450    0.620690  1.379310  0.123750    1.450000  
17               0.450    0.450    1.000000  2.222222  0.247500         inf  
18               0.725    0.350    1.000000  1.379310  0.096250         inf  
19               0.425    0.350    1.000000  2.352941  0.201250         inf  
20               0.350    0.350    0.823529  2.352941  0.201250    3.683333  
21               0.425    0.350    1.000000  2.352941  0.201250         inf  
22               0.350    0.350    0.823529  2.352941  0.201250    3.683333  
23               0.350    0.350    0.482759  1.379310  0.096250    1.256667

## Next Rule is according to Lift threshold where minimum Lift applied is for 2.1.
### Lift is a measure of correlation. Lift needs to be greater than 1 for the probability of 2 products to be bought together. For above dataset minimum Lift value is 1.3 and we are filetring for graeter than 2.1.

In [14]:
rules1=association_rules(freq_items1,metric="lift",min_threshold=2.1)
rules1

antecedents       consequents  antecedent support  \
0             (milo)         (borvita)               0.450   
1          (borvita)            (milo)               0.450   
2           (prtnex)         (complan)               0.350   
3          (complan)          (prtnex)               0.425   
4      (milo, holik)         (borvita)               0.450   
5   (holik, borvita)            (milo)               0.450   
6             (milo)  (holik, borvita)               0.450   
7          (borvita)     (milo, holik)               0.450   
8    (prtnex, holik)         (complan)               0.350   
9   (complan, holik)          (prtnex)               0.425   
10          (prtnex)  (complan, holik)               0.350   
11         (complan)   (prtnex, holik)               0.425   

    consequent support  support  confidence      lift  leverage  conviction  
0                0.450     0.45    1.000000  2.222222   0.24750         inf  
1                0.450     0.45    1.000000  2.222222   0.24750         inf  
2                0.425     0.35    1.000000  2.352941   0.20125         inf  
3                0.350     0.35    0.823529  2.352941   0.20125    3.683333  
4                0.450     0.45    1.000000  2.222222   0.24750         inf  
5                0.450     0.45    1.000000  2.222222   0.24750         inf  
6                0.450     0.45    1.000000  2.222222   0.24750         inf  
7                0.450     0.45    1.000000  2.222222   0.24750         inf  
8                0.425     0.35    1.000000  2.352941   0.20125         inf  
9                0.350     0.35    0.823529  2.352941   0.20125    3.683333  
10               0.425     0.35    1.000000  2.352941   0.20125         inf  
11               0.350     0.35    0.823529  2.352941   0.20125    3.683333

## Calculating length of Antecedents columns to check no. of products

In [15]:
rules1["antecedents_len"]=rules1["antecedents"].apply(lambda x:len(x))
rules1

antecedents       consequents  antecedent support  \
0             (milo)         (borvita)               0.450   
1          (borvita)            (milo)               0.450   
2           (prtnex)         (complan)               0.350   
3          (complan)          (prtnex)               0.425   
4      (milo, holik)         (borvita)               0.450   
5   (holik, borvita)            (milo)               0.450   
6             (milo)  (holik, borvita)               0.450   
7          (borvita)     (milo, holik)               0.450   
8    (prtnex, holik)         (complan)               0.350   
9   (complan, holik)          (prtnex)               0.425   
10          (prtnex)  (complan, holik)               0.350   
11         (complan)   (prtnex, holik)               0.425   

    consequent support  support  confidence      lift  leverage  conviction  \
0                0.450     0.45    1.000000  2.222222   0.24750         inf   
1                0.450     0.45    1.000000  2.222222   0.24750         inf   
2                0.425     0.35    1.000000  2.352941   0.20125         inf   
3                0.350     0.35    0.823529  2.352941   0.20125    3.683333   
4                0.450     0.45    1.000000  2.222222   0.24750         inf   
5                0.450     0.45    1.000000  2.222222   0.24750         inf   
6                0.450     0.45    1.000000  2.222222   0.24750         inf   
7                0.450     0.45    1.000000  2.222222   0.24750         inf   
8                0.425     0.35    1.000000  2.352941   0.20125         inf   
9                0.350     0.35    0.823529  2.352941   0.20125    3.683333   
10               0.425     0.35    1.000000  2.352941   0.20125         inf   
11               0.350     0.35    0.823529  2.352941   0.20125    3.683333   

    antecedents_len  
0                 1  
1                 1  
2                 1  
3                 1  
4                 2  
5                 2  
6                 1  
7                 1  
8                 2  
9                 2  
10                1  
11                1

## Final rule applied where Confidence is 100% and Lift is greater than 2.3
### Outcome:
This show that if a customer buys prtnex and holik, then there is 100% probability that they will buy compaln also as the Confidence is 1.0.

In [16]:
rules1[(rules1["antecedents_len"]>=2) & (rules1["confidence"]>=1) & (rules1["lift"]>2.3)]

antecedents consequents  antecedent support  consequent support  \
8  (prtnex, holik)   (complan)                0.35               0.425   

   support  confidence      lift  leverage  conviction  antecedents_len  
8     0.35         1.0  2.352941   0.20125         inf                2

## 2. Association Rule for Data for arules 2 :

In [17]:
arules2.head()

Row Labels  Almonds  Poha  Banana  Beans  Besan  Boiled Rice  Bread  \
0     6468572      NaN   NaN     1.0    NaN    NaN          NaN    NaN   
1     6486475      NaN   NaN     NaN    NaN    NaN          1.0    NaN   
2     6504964      1.0   NaN     NaN    NaN    NaN          1.0    NaN   
3     6529569      NaN   NaN     1.0    NaN    NaN          NaN    NaN   
4     6549521      NaN   NaN     1.0    NaN    NaN          NaN    NaN   

   Brinjals  Buns & Pavs  ...  Sooji & Rava  Soya Products  Sugar  \
0       NaN          NaN  ...           NaN            NaN    1.0   
1       NaN          1.0  ...           NaN            NaN    NaN   
2       NaN          NaN  ...           NaN            NaN    NaN   
3       NaN          NaN  ...           NaN            NaN    1.0   
4       NaN          NaN  ...           NaN            NaN    NaN   

   Sugar Cubes  Sunflower Oils  Toor Dal  Urad Dal  Utensil Scrub Pads  \
0          1.0             NaN       NaN       NaN                 NaN   
1          NaN             NaN       1.0       NaN                 1.0   
2          NaN             NaN       NaN       1.0                 NaN   
3          NaN             NaN       1.0       1.0                 NaN   
4          1.0             NaN       NaN       NaN                 NaN   

   Washing Bars  Whole Spices  
0           NaN           NaN  
1           NaN           NaN  
2           NaN           NaN  
3           NaN           NaN  
4           NaN           NaN  

[5 rows x 54 columns]

### Removing irrelevant column Row Labels 

In [18]:
newarules2 = pd.DataFrame(arules2)
newarules2.drop(["Row Labels"],axis=1,inplace=True)
newarules2.head()

Almonds  Poha  Banana  Beans  Besan  Boiled Rice  Bread  Brinjals  \
0      NaN   NaN     1.0    NaN    NaN          NaN    NaN       NaN   
1      NaN   NaN     NaN    NaN    NaN          1.0    NaN       NaN   
2      1.0   NaN     NaN    NaN    NaN          1.0    NaN       NaN   
3      NaN   NaN     1.0    NaN    NaN          NaN    NaN       NaN   
4      NaN   NaN     1.0    NaN    NaN          NaN    NaN       NaN   

   Buns & Pavs  Cakes  ...  Sooji & Rava  Soya Products  Sugar  Sugar Cubes  \
0          NaN    NaN  ...           NaN            NaN    1.0          1.0   
1          1.0    1.0  ...           NaN            NaN    NaN          NaN   
2          NaN    NaN  ...           NaN            NaN    NaN          NaN   
3          NaN    1.0  ...           NaN            NaN    1.0          NaN   
4          NaN    NaN  ...           NaN            NaN    NaN          1.0   

   Sunflower Oils  Toor Dal  Urad Dal  Utensil Scrub Pads  Washing Bars  \
0             NaN       NaN       NaN                 NaN           NaN   
1             NaN       1.0       NaN                 1.0           NaN   
2             NaN       NaN       1.0                 NaN           NaN   
3             NaN       1.0       1.0                 NaN           NaN   
4             NaN       NaN       NaN                 NaN           NaN   

   Whole Spices  
0           NaN  
1           NaN  
2           NaN  
3           NaN  
4           NaN  

[5 rows x 53 columns]

### Checking for Null values if any

In [19]:
newarules2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50 entries, 0 to 49
Data columns (total 53 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Almonds                     7 non-null      float64
 1   Poha                        2 non-null      float64
 2   Banana                      36 non-null     float64
 3   Beans                       24 non-null     float64
 4   Besan                       2 non-null      float64
 5   Boiled Rice                 20 non-null     float64
 6   Bread                       7 non-null      float64
 7   Brinjals                    2 non-null      float64
 8   Buns & Pavs                 1 non-null      float64
 9   Cakes                       6 non-null      float64
 10  Cashews                     10 non-null     float64
 11  Chips                       2 non-null      float64
 12  Cookies                     2 non-null      float64
 13  Corn Snacks                 4 non-nul

### Replaced Null values with 0 as Apriori alogorithm works on either 1 and 0 or True and False

In [20]:
newarules2.fillna(0,inplace=True)
newarules2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50 entries, 0 to 49
Data columns (total 53 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   Almonds                     50 non-null     float64
 1   Poha                        50 non-null     float64
 2   Banana                      50 non-null     float64
 3   Beans                       50 non-null     float64
 4   Besan                       50 non-null     float64
 5   Boiled Rice                 50 non-null     float64
 6   Bread                       50 non-null     float64
 7   Brinjals                    50 non-null     float64
 8   Buns & Pavs                 50 non-null     float64
 9   Cakes                       50 non-null     float64
 10  Cashews                     50 non-null     float64
 11  Chips                       50 non-null     float64
 12  Cookies                     50 non-null     float64
 13  Corn Snacks                 50 non-nu

In [21]:
newarules2.head()

Almonds  Poha  Banana  Beans  Besan  Boiled Rice  Bread  Brinjals  \
0      0.0   0.0     1.0    0.0    0.0          0.0    0.0       0.0   
1      0.0   0.0     0.0    0.0    0.0          1.0    0.0       0.0   
2      1.0   0.0     0.0    0.0    0.0          1.0    0.0       0.0   
3      0.0   0.0     1.0    0.0    0.0          0.0    0.0       0.0   
4      0.0   0.0     1.0    0.0    0.0          0.0    0.0       0.0   

   Buns & Pavs  Cakes  ...  Sooji & Rava  Soya Products  Sugar  Sugar Cubes  \
0          0.0    0.0  ...           0.0            0.0    1.0          1.0   
1          1.0    1.0  ...           0.0            0.0    0.0          0.0   
2          0.0    0.0  ...           0.0            0.0    0.0          0.0   
3          0.0    1.0  ...           0.0            0.0    1.0          0.0   
4          0.0    0.0  ...           0.0            0.0    0.0          1.0   

   Sunflower Oils  Toor Dal  Urad Dal  Utensil Scrub Pads  Washing Bars  \
0             0.0       0.0       0.0                 0.0           0.0   
1             0.0       1.0       0.0                 1.0           0.0   
2             0.0       0.0       1.0                 0.0           0.0   
3             0.0       1.0       1.0                 0.0           0.0   
4             0.0       0.0       0.0                 0.0           0.0   

   Whole Spices  
0           0.0  
1           0.0  
2           0.0  
3           0.0  
4           0.0  

[5 rows x 53 columns]

### Replacing values with data 2,3 and 4 with 1 

In [22]:
newarules2.replace([2,3,4],1,inplace=True)

In [23]:
newarules2.head()

Almonds  Poha  Banana  Beans  Besan  Boiled Rice  Bread  Brinjals  \
0      0.0   0.0     1.0    0.0    0.0          0.0    0.0       0.0   
1      0.0   0.0     0.0    0.0    0.0          1.0    0.0       0.0   
2      1.0   0.0     0.0    0.0    0.0          1.0    0.0       0.0   
3      0.0   0.0     1.0    0.0    0.0          0.0    0.0       0.0   
4      0.0   0.0     1.0    0.0    0.0          0.0    0.0       0.0   

   Buns & Pavs  Cakes  ...  Sooji & Rava  Soya Products  Sugar  Sugar Cubes  \
0          0.0    0.0  ...           0.0            0.0    1.0          1.0   
1          1.0    1.0  ...           0.0            0.0    0.0          0.0   
2          0.0    0.0  ...           0.0            0.0    0.0          0.0   
3          0.0    1.0  ...           0.0            0.0    1.0          0.0   
4          0.0    0.0  ...           0.0            0.0    0.0          1.0   

   Sunflower Oils  Toor Dal  Urad Dal  Utensil Scrub Pads  Washing Bars  \
0             0.0       0.0       0.0                 0.0           0.0   
1             0.0       1.0       0.0                 1.0           0.0   
2             0.0       0.0       1.0                 0.0           0.0   
3             0.0       1.0       1.0                 0.0           0.0   
4             0.0       0.0       0.0                 0.0           0.0   

   Whole Spices  
0           0.0  
1           0.0  
2           0.0  
3           0.0  
4           0.0  

[5 rows x 53 columns]

## Creating table for frequent products with minimum Support=30%

In [24]:
freq_items2=apriori(newarules2,min_support=0.3,use_colnames=True)
freq_items2

support                          itemsets
0      0.72                          (Banana)
1      0.48                           (Beans)
2      0.40                     (Boiled Rice)
3      0.42                         (Namkeen)
4      0.50                      (Other Dals)
5      0.40             (Other Rice Products)
6      0.38                (Other Vegetables)
7      0.30                         (Raisins)
8      0.44                 (Root Vegetables)
9      0.40                     (Snacky Nuts)
10     0.50                           (Sugar)
11     0.40                     (Sugar Cubes)
12     0.36                        (Urad Dal)
13     0.30                    (Whole Spices)
14     0.38                   (Banana, Beans)
15     0.30             (Boiled Rice, Banana)
16     0.38                 (Namkeen, Banana)
17     0.36              (Other Dals, Banana)
18     0.32        (Other Vegetables, Banana)
19     0.40         (Banana, Root Vegetables)
20     0.36             (Banana, Snacky Nuts)
21     0.38                   (Banana, Sugar)
22     0.32             (Banana, Sugar Cubes)
23     0.32               (Other Dals, Beans)
24     0.32         (Other Vegetables, Beans)
25     0.34          (Root Vegetables, Beans)
26     0.34                    (Sugar, Beans)
27     0.32         (Other Dals, Boiled Rice)
28     0.30     (Other Dals, Root Vegetables)
29     0.36               (Other Dals, Sugar)
30     0.30            (Urad Dal, Other Dals)
31     0.30          (Root Vegetables, Sugar)
32     0.34                 (Urad Dal, Sugar)
33     0.30  (Banana, Root Vegetables, Beans)

## Next Rule is applied for Confidence with minimum threshold of 50%

In [25]:
rules=association_rules(freq_items2,metric="confidence",min_threshold=0.5)
rules

antecedents                consequents  antecedent support  \
0                    (Banana)                    (Beans)                0.72   
1                     (Beans)                   (Banana)                0.48   
2               (Boiled Rice)                   (Banana)                0.40   
3                   (Namkeen)                   (Banana)                0.42   
4                    (Banana)                  (Namkeen)                0.72   
5                (Other Dals)                   (Banana)                0.50   
6                    (Banana)               (Other Dals)                0.72   
7          (Other Vegetables)                   (Banana)                0.38   
8                    (Banana)          (Root Vegetables)                0.72   
9           (Root Vegetables)                   (Banana)                0.44   
10                   (Banana)              (Snacky Nuts)                0.72   
11              (Snacky Nuts)                   (Banana)                0.40   
12                   (Banana)                    (Sugar)                0.72   
13                    (Sugar)                   (Banana)                0.50   
14              (Sugar Cubes)                   (Banana)                0.40   
15               (Other Dals)                    (Beans)                0.50   
16                    (Beans)               (Other Dals)                0.48   
17         (Other Vegetables)                    (Beans)                0.38   
18                    (Beans)         (Other Vegetables)                0.48   
19          (Root Vegetables)                    (Beans)                0.44   
20                    (Beans)          (Root Vegetables)                0.48   
21                    (Sugar)                    (Beans)                0.50   
22                    (Beans)                    (Sugar)                0.48   
23               (Other Dals)              (Boiled Rice)                0.50   
24              (Boiled Rice)               (Other Dals)                0.40   
25               (Other Dals)          (Root Vegetables)                0.50   
26          (Root Vegetables)               (Other Dals)                0.44   
27               (Other Dals)                    (Sugar)                0.50   
28                    (Sugar)               (Other Dals)                0.50   
29                 (Urad Dal)               (Other Dals)                0.36   
30               (Other Dals)                 (Urad Dal)                0.50   
31          (Root Vegetables)                    (Sugar)                0.44   
32                    (Sugar)          (Root Vegetables)                0.50   
33                 (Urad Dal)                    (Sugar)                0.36   
34                    (Sugar)                 (Urad Dal)                0.50   
35  (Banana, Root Vegetables)                    (Beans)                0.40   
36            (Banana, Beans)          (Root Vegetables)                0.38   
37   (Root Vegetables, Beans)                   (Banana)                0.34   
38          (Root Vegetables)            (Banana, Beans)                0.44   
39                    (Beans)  (Banana, Root Vegetables)                0.48   

    consequent support  support  confidence      lift  leverage  conviction  
0                 0.48     0.38    0.527778  1.099537    0.0344    1.101176  
1                 0.72     0.38    0.791667  1.099537    0.0344    1.344000  
2                 0.72     0.30    0.750000  1.041667    0.0120    1.120000  
3                 0.72     0.38    0.904762  1.256614    0.0776    2.940000  
4                 0.42     0.38    0.527778  1.256614    0.0776    1.228235  
5                 0.72     0.36    0.720000  1.000000    0.0000    1.000000  
6                 0.50     0.36    0.500000  1.000000    0.0000    1.000000  
7                 0.72     0.32    0.842105  1.169591    0.0464    1.773333  
8                 0.44     0.40    

## Next Rule applied for Lift with minimum threshold of 1.4.
### From below result it is clear which products are most popular with Lift>1

In [28]:
rules1=association_rules(freq_items2,metric="lift",min_threshold=1.6)
rules1

antecedents         consequents  antecedent support  \
0  (Other Vegetables)             (Beans)                0.38   
1             (Beans)  (Other Vegetables)                0.48   
2   (Root Vegetables)             (Beans)                0.44   
3             (Beans)   (Root Vegetables)                0.48   
4          (Urad Dal)        (Other Dals)                0.36   
5        (Other Dals)          (Urad Dal)                0.50   
6          (Urad Dal)             (Sugar)                0.36   
7             (Sugar)          (Urad Dal)                0.50   
8     (Banana, Beans)   (Root Vegetables)                0.38   
9   (Root Vegetables)     (Banana, Beans)                0.44   

   consequent support  support  confidence      lift  leverage  conviction  
0                0.48     0.32    0.842105  1.754386    0.1376    3.293333  
1                0.38     0.32    0.666667  1.754386    0.1376    1.860000  
2                0.48     0.34    0.772727  1.609848    0.1288    2.288000  
3                0.44     0.34    0.708333  1.609848    0.1288    1.920000  
4                0.50     0.30    0.833333  1.666667    0.1200    3.000000  
5                0.36     0.30    0.600000  1.666667    0.1200    1.600000  
6                0.50     0.34    0.944444  1.888889    0.1600    9.000000  
7                0.36     0.34    0.680000  1.888889    0.1600    2.000000  
8                0.44     0.30    0.789474  1.794258    0.1328    2.660000  
9                0.38     0.30    0.681818  1.794258    0.1328    1.948571

## Calculating the length of Antecedent column to check no. of products

In [29]:
rules1["antecedents_len"]=rules1["antecedents"].apply(lambda x:len(x))
rules1

antecedents         consequents  antecedent support  \
0  (Other Vegetables)             (Beans)                0.38   
1             (Beans)  (Other Vegetables)                0.48   
2   (Root Vegetables)             (Beans)                0.44   
3             (Beans)   (Root Vegetables)                0.48   
4          (Urad Dal)        (Other Dals)                0.36   
5        (Other Dals)          (Urad Dal)                0.50   
6          (Urad Dal)             (Sugar)                0.36   
7             (Sugar)          (Urad Dal)                0.50   
8     (Banana, Beans)   (Root Vegetables)                0.38   
9   (Root Vegetables)     (Banana, Beans)                0.44   

   consequent support  support  confidence      lift  leverage  conviction  \
0                0.48     0.32    0.842105  1.754386    0.1376    3.293333   
1                0.38     0.32    0.666667  1.754386    0.1376    1.860000   
2                0.48     0.34    0.772727  1.609848    0.1288    2.288000   
3                0.44     0.34    0.708333  1.609848    0.1288    1.920000   
4                0.50     0.30    0.833333  1.666667    0.1200    3.000000   
5                0.36     0.30    0.600000  1.666667    0.1200    1.600000   
6                0.50     0.34    0.944444  1.888889    0.1600    9.000000   
7                0.36     0.34    0.680000  1.888889    0.1600    2.000000   
8                0.44     0.30    0.789474  1.794258    0.1328    2.660000   
9                0.38     0.30    0.681818  1.794258    0.1328    1.948571   

   antecedents_len  
0                1  
1                1  
2                1  
3                1  
4                1  
5                1  
6                1  
7                1  
8                2  
9                1

## Final rule is applied for Confidence level 78% and Lift > 1.7
### Outcome:
Below result shows that if a customer buys Banana and Beans then there is 78.9% probability that they will buy Root Vegetables also as the Confidence is 78.9%. According to the analysis the products can be arranged in different patterns in store for maximum sale.

In [30]:
rules1[(rules1["antecedents_len"]>=2) & (rules1["confidence"]>=0.7) & (rules1["lift"]>1.7)]

antecedents        consequents  antecedent support  consequent support  \
8  (Banana, Beans)  (Root Vegetables)                0.38                0.44   

   support  confidence      lift  leverage  conviction  antecedents_len  
8      0.3    0.789474  1.794258    0.1328        2.66                2